src/services/local-storage/useLocalStorage.ts

```diff
@@ -1,32 +1,67 @@
-import type { Dispatch, SetStateAction } from 'react'
-import { useState, useCallback, useEffect } from 'react'
+import { useCallback, useEffect, useState } from 'react'
 import local from './local'
 
-const useLocalStorage = <T>(key: string, initialState: T): [T, Dispatch<SetStateAction<T>>] => {
-  const [cache, setCache] = useState<T>(initialState)
+// The setter accepts T or a function that takes the old value and returns T
+// Mimics the behavior of useState
+type Setter<T> = (val: T | ((prevVal: T) => T)) => void
 
-  useEffect(() => {
-    const initialValue = local.getItem<T>(key)
-    if (initialValue !== undefined) {
-      setCache(initialValue)
-    }
-  }, [setCache, key])
+const useLocalStorage = <T>(key: string, initialValue: T): [T, Setter<T>] => {
+  const [cache, setCache] = useState<T>(initialValue)
+
+  // This is the setter that will be returned
+  // It will update the local storage and cache
+  const setNewValue = useCallback<Setter<T>>(
+    (value) => {
+      setCache((oldValue) => {
+        const newValue = value instanceof Function ? value(oldValue) : value
 
-  const setNewValue = useCallback(
-    (value: T | ((prevState: T) => T)) => {
-      setCache((prevState) => {
-        const newState = value instanceof Function ? value(prevState) : value
+        if (newValue !== oldValue) {
+          local.setItem(key, newValue)
 
-        if (newState !== prevState) {
-          local.setItem(key, newState)
+          // Dispatch a fake storage event within the current browser tab
+          // The real storage event is dispatched only in other tabs
+          window.dispatchEvent(
+            new StorageEvent('storage', {
+              key: local.getPrefixedKey(key),
+            }),
+          )
         }
 
-        return newState
+        return newValue
       })
     },
-    [setCache, key],
+    [key],
   )
 
+  // Subscribe to changes in local storage and update the cache
+  // This will work across tabs
+  useEffect(() => {
+    const onStorageEvent = (event: StorageEvent) => {
+      if (event.key === local.getPrefixedKey(key)) {
+        const newValue = local.getItem<T>(key)
+        setCache((oldValue) => {
+          return newValue === undefined || newValue === oldValue ? oldValue : newValue
```

Comment:
Do we need to check anything here or can we simply return `newValue`?

For example if we use the hook twice for the same local storage key and one hook updates its value to `undefined`, should the other hook not also update its cache to `undefined`?

Reply:
The current usage of this hook unfortunately expects that the value is never undefined. I think it would be more correct if you could set it to undefined, but I kept the previous behavior.

Reply:
In safe creation we set the `pendingSafe` to `undefined` when the user cancels the transaction. With the change it is being set to `{}` which causes unintended behaviour. Maybe we could call `local.removeItem` when undefined is passed?

Reply:
OK, makes sense, let me allow undefined and remove the initialValue.